project/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -62,47 +70,72 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
-      Date date = (Date) election.getProperty("date");
-      List<String> candidateIds = (List<String>) election.getProperty("candidateIds");
-      List<String> candidateNames = new ArrayList<>(candidateIds.size());
-      List<String> candidatePositions = (List<String>) election.getProperty("candidatePositions");
-      List<String> candidatePartyAffiliation = new ArrayList<>(candidateIds.size());
-      List<String> candidateIncumbency= (List<String>) election.getProperty("candidateIncumbency");
-      // Query for a subset of candidate information.
-      for (String candidateId : candidateIds) {
-        Query candidateQuery = new Query("Candidate")
-            .setFilter(new FilterPredicate("__key__", FilterOperator.EQUAL,
-                       KeyFactory.createKey("Candidate", Long.parseLong(candidateId))));
-        PreparedQuery candidateQueryResult = datastore.prepare(candidateQuery);
-        Entity candidate = candidateQueryResult.asSingleEntity();
-        candidateNames.add((String) candidate.getProperty("name"));
-        candidatePartyAffiliation.add((String) candidate.getProperty("partyAffiliation"));
-      }
-      Election electionData = new Election(name, date, candidateIds, candidateNames,
-                                           candidatePositions, candidatePartyAffiliation,
-                                           candidateIncumbency);
-      electionsData.add(electionData);
+      Entity election = electionQueryResult.asSingleEntity();
+      List<DirectoryCandidate> candidates =
+          extractCandidateInformation((List<String>) election.getProperty("candidateIds"),
+                                      (List<String>) election.getProperty("candidateIncumbency"));
+      List<Position> positions =
+          extractPositionInformation((List<String>) election.getProperty("candidatePositions"),
+                                     candidates);
+      electionsData.add(new Election(election.getKey().getName(),
+                                     (Date) election.getProperty("date"),
+                                     positions));
     }
     return electionsData;
   }
 
+  /**
+   * Queries the database for (brief version) candidate information based on ID and formats the
+   * data as {@code DirectoryCandidate} objects.
+   */
+  private List<DirectoryCandidate> extractCandidateInformation(List<String> candidateIds,
+      List<String> candidateIncumbency) {
+    List<DirectoryCandidate> candidates = new ArrayList<>(candidateIds.size());
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    for (int i = 0; i < candidateIds.size(); i++) {
+      String id = candidateIds.get(i);
+      String incumbency = candidateIncumbency.get(i);
+      Query candidateQuery = new Query("Candidate")
+          .setFilter(new FilterPredicate("__key__", FilterOperator.EQUAL,
+                      KeyFactory.createKey("Candidate", Long.parseLong(id))));
```

Comment:
Is there a way to do a batch query? i.e. pass in the whole list of candidate IDs? Not sure if this is actually costing anything in the way of GCP credits (if not, feel free to ignore).

Reply:
It is, but it is a bit complex to execute. I'll look into this

Reply:
Datastore also supports an "IN" operator for a list of candidate IDs, in addition to "EQUAL". However, each individual ID will count as a query. "IN" also places a restriction that the size of the list must be < 30. This means that I would need to create sublists of size 30. Therefore, this approach probably won't save queries or simplify the code.

Reply:
There is also a [bulk deletion pipeline](https://cloud.google.com/dataflow/docs/guides/templates/provided-utilities#datastore-bulk-delete). However, this doesn't fit our particular use case